lib/inputs/dovetail-impressions.js

```diff
@@ -0,0 +1,92 @@
+'use strict';
+
+const crypto = require('crypto');
+const timestamp = require('../timestamp');
+const lookip = require('../lookip');
+const lookagent = require('../lookagent');
+const bigquery = require('../bigquery');
+
+/**
+ * Send dovetail impressions to bigquery
+ */
+module.exports = class DovetailImpressions {
+
+  constructor(records) {
+    this._records = (records || []).filter(r => this.check(r));
+  }
+
+  check(record) {
+    return (record.type === undefined && !!record.adId) || record.type === 'impression';
+  }
+
+  insert() {
+    if (this._records.length == 0) {
+      return Promise.resolve([]);
+    } else {
+      return Promise.all(this._records.map(r => this.format(r))).then(formats => {
+        let tables = {};
+        formats.forEach(f => tables[f.table] = (tables[f.table] || []).concat(f.record));
+
+        // run per-table inserts in parallel
+        return Promise.all(Object.keys(tables).map(t => {
+          return bigquery.insert(t, tables[t]).then(num => {
+            return {count: num, dest: t};
+          });
+        }));
+      });
+    }
+  }
+
+  format(record) {
+    let lookups = [lookip.look(record.remoteIp), lookagent.look(record.remoteAgent)];
+    return Promise.all(lookups).then(([geo, agent]) => {
+      return {
+        table: this.table(record),
+        record: {
+          // TODO: this ASSUMES any ad is only found once per arrangement!!!
```

Comment:
This is generally the case, but we should make sure some level of "Don't allow duplicates" is enabled, at least at the creative/ad level, to prevent this. Worth mentioning to Kali?

Reply:
Definitely.  And to be clear ... Adzerk would get the dups.  It's just our BQ table would have a lower impression count.

Reply:
yeah, really want to keep those in synch (as much as we can)